jetty-http/src/main/java/org/eclipse/jetty/http/HttpMethod.java

```diff
@@ -26,162 +26,179 @@
-     */
-    public static HttpMethod lookAheadGet(ByteBuffer buffer)
+    HttpMethod(String method, Type type)
     {
-        if (buffer.hasArray())
-            return lookAheadGet(buffer.array(), buffer.arrayOffset() + buffer.position(), buffer.arrayOffset() + buffer.limit());
-
-        int l = buffer.remaining();
-        if (l >= 4)
-        {
-            HttpMethod m = CACHE.getBest(buffer, 0, l);
-            if (m != null)
-            {
-                int ml = m.asString().length();
-                if (l > ml && buffer.get(buffer.position() + ml) == ' ')
-                    return m;
-            }
-        }
-        return null;
+        _method = method;
+        _type = type;
+        _bytes = StringUtil.getBytes(_method);
+        _buffer = ByteBuffer.wrap(_bytes);
     }
 
-    public static final Trie<HttpMethod> INSENSITIVE_CACHE = new ArrayTrie<>();
+    public byte[] getBytes()
+    {
+        return _bytes;
+    }
 
-    static
+    public boolean is(String s)
     {
-        for (HttpMethod method : HttpMethod.values())
-        {
-            INSENSITIVE_CACHE.put(method.toString(), method);
-        }
+        return toString().equalsIgnoreCase(s);
     }
 
-    public static final Trie<HttpMethod> CACHE = new ArrayTernaryTrie<>(false);
+    /**
+     * An HTTP method is safe if it doesn't alter the state of the server.
+     * In other words, a method is safe if it leads to a read-only operation.
+     * Several common HTTP methods are safe: GET , HEAD , or OPTIONS .
+     * All safe methods are also idempotent, but not all idempotent methods are safe
+     * @return if the method is safe.
+     */
+    public boolean isSafe()
+    {
+        return _type == Type.SAFE;
+    }
 
-    static
+    /**
+     * An idempotent HTTP method is an HTTP method that can be called many times without different outcomes.
+     * It would not matter if the method is called only once, or ten times over. The result should be the same.
+     * @return true if the method is idempotent.
+     */
+    public boolean isIdempotent()
     {
-        for (HttpMethod method : HttpMethod.values())
-        {
-            CACHE.put(method.toString(), method);
-        }
+        return _type.ordinal() >= Type.IDEMPOTENT.ordinal();
     }
 
-    private final ByteBuffer _buffer;
-    private final byte[] _bytes;
+    public ByteBuffer asBuffer()
+    {
+        return _buffer.asReadOnlyBuffer();
+    }
 
-    HttpMethod()
+    public String asString()
     {
-        _bytes = StringUtil.getBytes(toString());
-        _buffer = ByteBuffer.wrap(_bytes);
+        return _method;
     }
 
-    public byte[] getBytes()
+    public String toString()
     {
-        return _bytes;
+        return _method;
     }
 
-    public boolean is(String s)
+    public static final Trie<HttpMethod> INSENSITIVE_CACHE = new ArrayTrie<>(252);
+    public static final Trie<HttpMethod> CACHE = new ArrayTernaryTrie<>(false, 300);
```

Comment:
This cache is used by `HPackContext`, and you changed it from case-insensitive to case-sensitive. Was that on purpose?

Reply:
oops no.

Reply:
Let me add a test to check I don't do that again....

Reply:
No I didn't!!! the insensitive cache is `INSENSITIVE_CACHE`. The `CACHE` is correctly passed false for insensitive!